### ISSUE TITLE
Add Goal Oriented Action Planning (Goap)

### ISSUE THREAD

**Issue by hidemat** [godotengine/godot-proposals#4954]
### Describe the project you are working on

An implementation of Goal Oriented Action Planning (Goap) for managing complex AI behavior.

### Describe the problem or limitation you are having in your project

The Goap system is very complex to implement.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Provide ready to use classes and nodes for easier goap implementation

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Here is are some resources describing how Goap works:
[https://alumni.media.mit.edu/~jorkin/goap.html](https://alumni.media.mit.edu/~jorkin/goap.html)
https://www.youtube.com/watch?v=PaOLBOuyswI&t=1s&ab_channel=AIandGames

I'm hoping to provide custom nodes and resources that encompass what is proposed in the article.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It's often the case where complex AI is needed, but implementation in GDscript can be cumbersome. So an integrated solution would be better.

### Is there a reason why this should be core and not an add-on in the asset library?

I think it's a very common need to have a Goap system, but it's extremely hard to learn for novice programmers. Having a core goap system, that is well documented, would be a plus to all godot developers, in my opinion. I'm actually already working on a c++ module implementation for Godot 4. All I need is feedback to see if the community wants this.

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot-proposals/issues/281.

**Comment by fire** [comment]
I have a test for htn but was stuck exposing the api. 

https://github.com/V-Sekai/godot_hierarchical_task_network

**Comment by fire** [comment]
Your proposed project section isn’t wrll written, but I wanted to use this for npcs and companions / enemies.

**Comment by and-rad** [comment]
I think I am against this. What makes GOAP more generic / easier to use / better than, let's say, behavior trees or Utility AI? A [very good case](https://m.youtube.com/watch?v=Qq_xX1JCreI) can be made against using behavior trees for decision making, but an AI system will always be strongly taylored to the specific type of game that's being created. I believe a plugin would be better suited for this.

**Comment by Calinou** [comment]
I'd strongly recommend experimenting with this as an add-on, and see if anyone needs to be made generic for performance reasons or similar.

**Comment by hidemat** [comment]
Will do. I'm building this for a personal project of mine. I have invested a great deal of time learning how Goap works, so for me, this is the system I'm going to use. I just thought other people might what access to it as well.
So I'll finish writing it and make available for the community to play around with it.

**Comment by hidemat** [comment]
> I think I am against this. What makes GOAP more generic / easier to use / better than, let's say, behavior trees or Utility AI? A [very good case](https://m.youtube.com/watch?v=Qq_xX1JCreI) can be made against using behavior trees for decision making, but an AI system will always be strongly taylored to the specific type of game that's being created. I believe a plugin would be better suited for this.

**Comment by hidemat** [comment]
Sorry wrong button.

**Comment by barbaros83** [comment]
are you going to make this available as a gdextension ?

**Comment by hidemat** [comment]
Yeah. I actually got it working as a module when I posted this, but the code got broken at some point. I'm just trying to figure out how to port it to GDExtention.

**Comment by fire** [comment]
You can also see my unit tests at https://github.com/V-Sekai/godot-task-goal-planner
